pretrain/tinyllama.py

```diff
@@ -0,0 +1,369 @@
+            time=(time.time() - total_t0), 
+            samples=((state["iter_num"] + 1) * micro_batch_size), 
+            flops_per_batch=measured_flops,
+            lengths=total_lengths,
+        )
+
+        if state["iter_num"] % log_iter_interval == 0:
+            loss = loss.item()
+            t1 = time.perf_counter()
+            metrics = {
+                "loss": loss,
+                "iter": state['iter_num'],
+                "step": state['step_count'],
+                "iter_time": (t1 - iter_t0),
+                "remaining_time": (t1 - total_t0) / (state['iter_num'] - initial_iter) * (max_iters - state['iter_num']),
+            }
+
+            fabric.print(
+                f"iter {metrics['iter']} step {metrics['step']}: loss {metrics['loss']:.4f}, iter time:"
+                f" {metrics['iter_time'] * 1000:.2f}, ms{' (optimizer.step)' if not is_accumulating else ''}"
+                f" remaining time: {metrics['remaining_time'] / 3600 / 24:.2f} days"
+            )
+
+            throughput_metrics = throughput.compute()
+            print(state["iter_num"], throughput_metrics)
+            metrics.update(throughput_metrics)
+            fabric.log_dict(metrics)
+
+        if val_dataloader is not None and not is_accumulating and state["step_count"] % eval_step_interval == 0:
+            t0 = time.perf_counter()
+            val_loss = validate(fabric, model, val_dataloader)
+            t1 = time.perf_counter() - t0
+
+            fabric.print(f"step {state['iter_num']}: val loss {val_loss:.4f}, val time: {t1 * 1000:.2f}ms")
+            fabric.log_dict({"metric/val_loss": val_loss.item(), "total_tokens":  model.config.block_size * (state["iter_num"] + 1) * micro_batch_size * fabric.world_size},state["step_count"])
+            fabric.log_dict({"metric/val_ppl": math.exp(val_loss.item()), "total_tokens":  model.config.block_size * (state["iter_num"] + 1) * micro_batch_size * fabric.world_size},state["step_count"])
+            fabric.barrier()
+
+        if not is_accumulating and state["step_count"] % save_step_interval == 0:
+            checkpoint_path = out_dir / f"iter-{state['iter_num']:06d}-ckpt.pth"
+            fabric.print(f"Saving checkpoint to {str(checkpoint_path)!r}")
+            fabric.save(checkpoint_path, state)
+
+
+@torch.no_grad()
+def validate(fabric: L.Fabric, model: nn.Module, val_dataloader: DataLoader) -> torch.Tensor:
+    fabric.print("Validating ...")
+    model.eval()
+
+    losses = torch.zeros(eval_iters, device=fabric.device)
+    for k, val_data in enumerate(val_dataloader):
+        if k >= eval_iters:
+            break
+        input_ids = val_data[:, 0:model.config.block_size].contiguous().long()
+        targets = val_data[:, 1:(model.config.block_size + 1)].contiguous().long()
+        logits = model(input_ids)
+        loss = chunked_cross_entropy(logits, targets, chunk_size=0)
+        losses[k] = loss.item()
+
+    out = losses.mean()
+    model.train()
+    return out
```

Comment:
This doesn't need to `.item()` here. It can be done in the calls that use it

```suggestion
        losses[k] = loss

    out = losses.mean()
    model.train()
    return out
```